### PR TITLE
Add Logging for connections

### PR DESCRIPTION
**Pull request by legionth**
This PRs enables logging of the client that connect to LeProxy. This will be automatically started when LeProxy is started.

### CODE REVIEW

**Comment by legionth** [comment]
This is how the output of the log looks like:
```
$ php leproxy.php 
LeProxy HTTP/SOCKS proxy now listening on http://127.0.0.1:1080
2017-06-14 10:36:44 connected tcp://127.0.0.1:38912
2017-06-14 10:36:44 disconnected tcp://127.0.0.1:38912
2017-06-14 10:36:44 connected tcp://127.0.0.1:38916
2017-06-14 10:36:44 disconnected tcp://127.0.0.1:38916
2017-06-14 10:36:46 connected tcp://127.0.0.1:38922
2017-06-14 10:36:46 connected tcp://127.0.0.1:38926
2017-06-14 10:36:47 disconnected tcp://127.0.0.1:38926
2017-06-14 10:36:47 connected tcp://127.0.0.1:38930
```

File: src/LoggingConnector.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+use React\Socket\Connector;
+use React\Promise\Promise;
+use React\Stream\WritableStreamInterface;
+use React\Socket\ConnectorInterface;
+use React\Socket\ConnectionInterface;
+use Clue\React\HttpProxy\ProxyConnector as HttpClient;
+use Clue\React\Socks\Client as SocksClient;
+
+/** @internal */
```
Comment (clue) [inline src/LoggingConnector.php]: I think everything is "internal" here :-) Perhaps add some (concise) documentation about what this class does instead?

File: src/LoggingConnector.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    public function __construct(ConnectorInterface $connector, WritableStreamInterface $outputStream)
+    {
+        if ($connector instanceof HttpClient) {
+            $this->protocol = 'http';
+        }
+
+        if ($connector instanceof SocksClient) {
+            $this->protocol = 'socks5';
+        }
```
Comment (clue) [inline src/LoggingConnector.php]: Not a big fan of these checks here, also doesn't look right. This check only ensures a connection from LeProxy to the next upstream proxy will be through "socks5". This is not to be confused with the protocol where this connection came *in* from.

File: src/LoggingConnector.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+                        $that->log($message);
+                        $resolve($remote);
+                    }
+                );
+            }
+        );
+    }
+
+    public function log($message)
```
Comment (clue) [inline src/LoggingConnector.php]: Does it make sense to keep this public? (see also others)

File: src/LoggingConnector.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    public function connect($uri)
+    {
+        $that = $this;
+        $connector = $this->connector;
+        $auth = $this->auth;
+
+        return new Promise(
+            function ($resolve, $reject) use ($connector, $uri, $that, $auth) {
+                $connector->connect($uri)->then(
+                    function (ConnectionInterface $remote) use ($that, $uri, $resolve, $auth) {
```
Comment (clue) [inline src/LoggingConnector.php]: Code style is a bit unclear here, does this target PHP < 5.4, in which case `$this` must not be used or if this targets only PHP > 5.3, then most of these variable bindings are unneeded.

File: src/LoggingConnector.php [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+    public function log($message)
+    {
+        $this->outputStream->write(date('Y-m-d H:i:s') . ' ' . $message. PHP_EOL);
+    }
+
+    public function setAuth($auth)
+    {
+        $this->auth = $auth;
+    }
```
Comment (clue) [inline src/LoggingConnector.php]: Looks a bit unclear API-wise, what goes in, what goes out and why? Looks like an array is being passed here, while the above logic doesn't really apply to multiple entries, no?

File: tests/LoggingConnectorTest.php [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        $this->stream->on('data', function($data) use (&$result) {
+            $result .= $data;
+        });
+
+        $logging = new LoggingConnector($this->httpConnector, $this->stream);
+        $logging->setAuth(array('legionth' => 'test'));
+
+        $logging->connect('google.com:443');
+
+        $this->assertContains('connected legionth@http://192.168.1.2 to tcp://google.com:443', $result);
```
Comment (clue) [inline tests/LoggingConnectorTest.php]: This is an *interesting* format, does it make sense to use standard URI format here, à la `http://legionth@192.168.1.2`?

**Comment by clue** [comment]
Thanks for the PR! I feel most of my above comments can be addressed quite easily, but I don't think *code* is really an issue here.

Instead, I'd rather talk *specs* here and instead ask *what* should be implemented as part of this PR?

http://thecodinglove.com/post/162435568195/delivering-a-feature-as-described-in-specs

What is *logging*, what is the *use case*, what *should* be logged, how is this presented?

It's my understanding that this is currently meant as a tool to help diagnose operation of LeProxy. In this case, it probably makes little sense to *only* log successful connections, as this would leave the user  clueless as to why certain sites do not work?

Also, I don't see any mention of this feature in the README and *why* this feature exists in the first place (motivation unclear).

What do you think?

**Pull request closed without merging**